pkg/controllermanager/controller/managedseedset/managedseedset_actuator.go

```diff
@@ -0,0 +1,414 @@
+		}
+		return true, nil
+
+	case replicaStatus == StatusShootReconciling && !scalingIn:
+		// This replica's shoot is reconciling, wait for it to be reconciled before moving to the next replica
+		a.infoEventf(set, "Waiting for shoot %s to be reconciled", r.GetFullName())
+		updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ShootReconcilingReason, nil)
+		return true, nil
+
+	case replicaStatus == StatusShootDeleting:
+		// This replica's shoot is deleting, wait for it to be deleted before moving to the next replica
+		a.infoEventf(set, "Waiting for shoot %s to be deleted", r.GetFullName())
+		updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ShootDeletingReason, nil)
+		return true, nil
+
+	case replicaStatus == StatusShootReconciled:
+		// This replica's shoot is fully reconciled and its managed seed doesn't exist
+		// If not scaling in, create its managed seed, otherwise delete its shoot
+		if !scalingIn {
+			a.infoEventf(set, "Creating managed seed %s", r.GetFullName())
+			if err := r.CreateManagedSeed(ctx, a.gardenClient.Client()); err != nil {
+				return false, err
+			}
+			updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ManagedSeedPreparingReason, nil)
+		} else {
+			a.infoEventf(set, "Deleting shoot %s", r.GetFullName())
+			if err := r.DeleteShoot(ctx, a.gardenClient.Client()); err != nil {
+				return false, err
+			}
+			updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ShootDeletingReason, nil)
+		}
+		return true, nil
+
+	case replicaStatus == StatusManagedSeedPreparing && !scalingIn:
+		// This replica's managed seed is preparing, wait for the it to be registered before moving to the next replica
+		a.infoEventf(set, "Waiting for managed seed %s to be registered", r.GetFullName())
+		updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ManagedSeedPreparingReason, nil)
+		return true, nil
+
+	case replicaStatus == StatusManagedSeedDeleting:
+		// This replica's managed seed is deleting, wait for it to be deleted before moving to the next replica
+		a.infoEventf(set, "Waiting for managed seed %s to be deleted", r.GetFullName())
+		updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ManagedSeedDeletingReason, nil)
+		return true, nil
+
+	case !r.IsSeedReady() && !scalingIn:
+		// This replica's seed is not ready, wait for it to be ready before moving to the next replica
+		a.infoEventf(set, "Waiting for seed %s to be ready", r.GetName())
+		updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.SeedNotReadyReason, nil)
+		return true, nil
+
+	case r.GetShootHealthStatus() != operationshoot.StatusHealthy && !scalingIn:
+		// This replica's shoot is not healthy, wait for it to be healthy before moving to the next replica
+		a.infoEventf(set, "Waiting for shoot %s to be healthy", r.GetFullName())
+		updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ShootNotHealthyReason, nil)
```

Comment:
If a set is currently ready (all replicas healthy and no pending replica) and one of the replicas becomes unhealthy, we make it the pending replica and wait for it to be created.
Now that we ensure, that we don't block ongoing scale operations when previous replicas become unhealthy, I think we should similarly ensure, that a transient/flaky replica health status doesn't block future scale operations. I.e. we should probably not set the pending replica if a shoot's health status changes and only update the `readyReplicas` field.

Reply:
I had similar thoughts, but there is a problem: it's from my perspective impossible to properly distinguish between 1) a shoot that was never healthy after it was created and 2) a shoot that just became unhealthy. Consider for example the last replica during a scale-out: since the target replica count has been reached, if this replica is unhealthy it's impossible to tell in which of the 2 situations we are in.

I think in the first case, we definitely want to stop and wait until it becomes healthy, since the reason might well be in our MSS resource.

My personal opinion is that we shouldn't try to make this too perfect. I would err on the "safe side", and let future scale-out operations be blocked in cases like this. If they are blocked for too long, likely the issue is going to prevent the proper scale-out anyway. Note that only scale-outs will be blocked, scale-ins will still be performed because there the shoot health is not taken into account.